recipes-ni/pstore-save/files/pstore-save

```diff
@@ -0,0 +1,139 @@
+#!/bin/bash
+
+# Copyright (c) 2021 National Instruments
+#
+# SPDX-License-Identifier: MIT
+#
+# This script retrieves data from "pstore", which is a persistent
+# storage interface that the kernel can use to store "last breath"
+# crash information. On x64 targets, this takes the form of EFI
+# variables. Because there is limited space available in pstore,
+# this script is intended to run on boot to archive such information
+# to disk.
+#
+# By default, dmesg information from last boot is reconstructed in
+#    /var/lib/pstore/YYYYMMDD_hhmmss/dmesg-nnn.txt
+#
+# This script is largely inspired by (and obsoleted by) systemd-pstore,
+# added in systemd 243.
+
+set -euo pipefail
+
+SCRIPTNAME=$(basename "$0")
+
+UNLINK=1
+SOURCEDIR="/sys/fs/pstore"
+ARCHIVEDIR="/var/lib/pstore"
+
+function print_usage_and_die () {
+    cat >&2 <<ENDUSAGE
+$0 [-a archivedir] [-k] [-s sourcedir] [-d]
+   -a     output directory for pstore data ($ARCHIVEDIR)
+   -k     keep pstore data (default: remove from pstore)
+   -s     source directory ($SOURCEDIR)
+ENDUSAGE
+    exit 1
+}
+
+while getopts "ha:ks:" arg; do
+    case ${arg} in
+    a)  ARCHIVEDIR=${OPTARG} ;;
+    k)  UNLINK=0 ;;
+    s)  SOURCEDIR=${OPTARG} ;;
+    h|?) print_usage_and_die ;;
+    esac
+done
+
+function log_msg () {
+    echo "$SCRIPTNAME: $@"
+}
+
+function log_error () {
+    echo "$SCRIPTNAME: $@" >&2
+}
+
+# move_file <src> <dst-subdir>
+function move_file () {
+    if ! mkdir -p -m 0755 "${ARCHIVEDIR}/${2}"; then
+        log_error "Failed to create ${ARCHIVEDIR}/${2}"
+        return 1
+    fi
+
+    if ! cp -a "$1" "${ARCHIVEDIR}/${2}"; then
+        log_error "Failed to copy ${1} to ${ARCHIVEDIR}/${2}"
+        return 1
+    fi
+
+    if [ $UNLINK -ne 0 ]; then
+        rm "$1" || true
+    fi
+}
+
+function check_for_mount () {
+    # If the source directory is /sys/fs/pstore, make sure that
+    # it has been mounted, and mount it if it has not. (If the
+    # source directory is not /sys/fs/pstore, you're on your own.)
+    if [ "${SOURCEDIR}" = "/sys/fs/pstore" ]; then
+        # check /proc/mtab for mounts
+        if ! grep "/sys/fs/pstore" "/proc/mounts"; then
+            if ! mount -t pstore pstore "/sys/fs/pstore"; then
+                log_error "/sys/fs/pstore wasn't mounted and can't be mounted"
+            fi
+        fi
+    fi
+}
+
+function process_dmesg_files () {
+    declare -A LOGGED_TIMESTAMPS
+
+    # Any files that start with "dmesg" and end in ".enc.z" are bad;
+    # move them to the archive and don't try to process them otherwise.
```

Comment:
Could you clarify this comment? What does "are bad" mean, in this context?

Reply:
I explain it about [as well as systemd-pstore does](https://github.com/systemd/systemd/blob/df8fd03a7dcf2f06cd9e2ee3eed60c3a83b43ccc/src/pstore/pstore.c#L229). :) I can update the comment to clarify though.

In practice, it seems to mean "the underlying pstore backend says that these are compressed data elements, but an error occurred when the backend attempted to decompress them in order to present them as entries in /sys/fs/pstore (where from the point-of-view of the reader they are uncompressed text files)".

In the case of the "nicleanefivars is deleting the dump variables out from behind the back of pstore" case, the way that this would manifest is that the /sys/fs/pstore directory would be full of a bunch of 1024-bytes-of-zeroes files named `dmesg-efi-157349851401001.enc.z` (and also get some dmesg spew of `pstore: decompression failed: -5`)